src/QMCDrivers/DMC/WalkerControlMPI.cpp

```diff
@@ -390,4 +447,163 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
   }
 }
 
+/** swap Walkers between rank MCPopulations
+ *
+ *  2005 called and asked for its synchronous messaging back
+ *  I think MPI async is more than two years old
+ *
+ *  MCPopulation is sufficiently different from MCWalkerConfiguration that this is 
+ *  basically a rewrite.  I have not replicated the more ridiculous elements of the original.
+ *  For example:
+ *   * Sending duplicates to the same reciever rank but resorting good walkers if send-receive pair 
+ *     changes.
+ *   * Sending a tiny separate blocking message to tell the receiver how many copies
+ *
+ */
+void WalkerControlMPI::swapWalkersSimple(MCPopulation& pop, PopulationAdjustment& adjust)
+{
+  std::vector<int> minus, plus;
+  determineNewWalkerPopulation(Cur_pop, num_contexts_, MyContext, NumPerNode, FairOffSet, minus, plus);
+
+  if (adjust.good_walkers.empty() && adjust.bad_walkers.empty())
+  {
+    app_error() << "It should never happen that no walkers, "
+                << "neither good nor bad, exist on a node. "
+                << "Please report to developers. " << std::endl;
+    APP_ABORT("WalkerControlMPI::swapWalkersSimple no existing walker");
+  }
+
+  if (plus.size() != minus.size())
+  {
+    app_error() << "Walker send/recv pattern doesn't match. "
+                << "The send size " << plus.size() << " is not equal to the receive size " << minus.size() << " ."
+                << std::endl;
+    throw std::runtime_error("Trying to swap in WalkerControlMPI::swapWalkersSimple with mismatched queues");
+  }
+
+  UPtrVector<MCPWalker> new_walkers;
+  std::vector<int> copies_of_each_new_walker;
+
+  // sort good walkers by the number of copies
+  std::vector<std::pair<int, std::reference_wrapper<MCPWalker>>> sorted_good_walkers;
+  for (int iw = 0; iw < ncopy_w.size(); iw++)
+    sorted_good_walkers.push_back(std::make_pair(adjust.copies_to_make[iw], adjust.good_walkers[iw]));
+  // Sort only on the number of copies
+  std::sort(sorted_good_walkers.begin(), sorted_good_walkers.end(),
+            [](auto& a, auto& b) { return a.first > b.first; });
```

Comment:
I thought std::sort works with vector pairs by the first value.

Reply:
https://en.cppreference.com/w/cpp/utility/pair/operator_cmp
TLDR
First compare lhs.first rhs.first if lhs.first == rhs.first then compare lhs.second, rhs.second.

I considered this oversorting since from my naive perspective sorting by walker IDs seemed if anything a potential bias.
However as concurrency, threaded and MPI, cause divergence this could add to it.

So while subtle this is a change.